jOOL-java-8/src/main/java/org/jooq/lambda/tuple/Tuple2.java

```diff
@@ -237,19 +229,39 @@ public final Tuple0 skip2() {
     }
 
     /**
-     * Whether two tuples overlap.
+     * Whether two tuples overlap. Assumes inclusiveness.
      * <p>
      * <code><pre>
      * // true
      * range(1, 3).overlaps(range(2, 4))
+     * range(1, null).overlaps(range(null, 1))
      *
      * // false
      * range(1, 3).overlaps(range(5, 8))
+     * range(null, 3).overlaps(range(5, null))
      * </pre></code>
      */
     public static final <T extends Comparable<? super T>> boolean overlaps(Tuple2<T, T> left, Tuple2<T, T> right) {
-        return left.v1.compareTo(right.v2) <= 0
-            && left.v2.compareTo(right.v1) >= 0;
+        if (left == null || right == null) {
+            return false;
+        }
+
+        if (left.infinite() || right.infinite()) {
+            return true;
+        }
+        if (left.finite() && right.finite()) {
+            return left.v1.compareTo(right.v2) <= 0
+                    && left.v2.compareTo(right.v1) >= 0;
+        }
+
+        if (!left.upperInfinite() && !right.lowerInfinite()) {
+            return left.v2.compareTo(right.v1) >= 0;
+        }
+        if (!left.lowerInfinite() && !right.upperInfinite()) {
+            return left.v1.compareTo(right.v2) <= 0;
+        }
+        // two remaining positive cases: either both have upper or lower bound infinite
+        return true;
```

Comment:
This seems much more concise:

```java
    public boolean overlaps(Range<T> other) {
        return (v1 == null
            ||  other.v2 == null
            ||  v1.compareTo(other.v2) <= 0)
            && (v2 == null
            ||  other.v1 == null
            ||  v2.compareTo(other.v1) >= 0);
    }
```

Reply:
Yes but at the cost of readability.
But readability is a bit subjective of course.

Reply:
Sure, we can use subjectivity here.

I find your version much less readable, because it takes me much longer to digest. For example, why is `left.infinite() || right.infinite()` a special case worth short circuiting? Why is the `left.finite() && right.finite()` version short circuited right after? Why do we check for infinity several times?